engine/src/actions/movement.rs

```diff
@@ -36,13 +38,38 @@ pub fn down(entity_id: i32) -> ActionData {
     }
 }
 
-pub fn process(world: &mut World, action: ActionData) {
-    if let Some(entity) = world.get_entity(action.entity_id) {
-        if let Some(dir) = action.direction {
-            let coord = operate(entity.coord, dir);
-            let new_entity = entity.with_coordinate(coord);
-            world.update_entity(new_entity);
-        }
+pub fn process(world: &mut World, action: ActionData) -> EngineResult<()> {
+    let entity = world.get_entity(action.entity_id)?;
+
+    if let Some(dir) = action.direction {
+        let new_coord = operate(entity.coord, dir);
+
+        is_inside_world(world, new_coord)?;
+        is_position_available(world, new_coord)?;
+
+        let new_entity = entity.with_coordinate(new_coord);
+        world.update_entity(new_entity);
+    }
+
+    Ok(())
+}
+
+fn is_inside_world(world: &World, coord: Coordinate) -> EngineResult<()> {
+    if coord.x < 0.0
+        || coord.y < 0.0
+        || coord.x >= world.size_x as f32
+        || coord.y >= world.size_y as f32
+    {
+        return Err(EngineError::OutOfMapCoordinate(coord.x, coord.y));
+    }
+
+    Ok(())
```

Comment:
Small style detail; feel free to ignore. I'd favor `if { ... } else { ... }` instead of using the explicit return statement here.

Reply:
That's interesting. I usually favour having guard clauses. Maybe because rubocop is just a PITA about it and I got used to it.

Reply:
If you notice, there's somewhere else in the PR where you used the same pattern and I didn't recommend changing it. The rule that I use is using `if-else` in cases like this, in which there's a couple of lines per branch, or branches have the same weight in the function definition.

I save guard clauses to checks that are like prerequisites to the execution of the main goal of the function, something that needs to be checked at the beginning. I think even the term "guard clause" hints at that kind of usage.

I hope that makes sense.

Reply:
Makes sense 👍